vlog4j-graal/src/main/java/org/semanticweb/vlog4j/graal/GraalImporter.java

```diff
@@ -0,0 +1,214 @@
+		return result;
+	}
+
+	/**
+	 * Converts a {@link AtomSet Graal AtomSet} into a {@link Conjunction VLog4J Conjunction}.
+	 * @param atomSet A {@link AtomSet Graal Atomset}
+	 * @return A {@link Conjunction VLog4J Conjunction}
+	 */
+	private static Conjunction importAtomSet(AtomSet atomSet) {
+		List<Atom> result = new ArrayList<>();
+		try (CloseableIterator<fr.lirmm.graphik.graal.api.core.Atom> iterator = atomSet.iterator()) {
+			while (iterator.hasNext())
+				result.add(importAtom(iterator.next()));
+		} catch (IteratorException e) {
+			throw new GraalImportException(MessageFormat.format("Unexpected Iterator Exception when importing AtomSet {0}}.", atomSet));
+		}
+		return makeConjunction(result);
+	}
+
+	/**
+	 * Converts a {@link fr.lirmm.graphik.graal.api.core.Constant Graal Constant} into a {@link Constant VLog4J Constant}.
+	 * @param constant A {@link fr.lirmm.graphik.graal.api.core.Constant Graal Constant}
+	 * @return A {@link Constant VLog4J Constant}
+	 */
+	private static Constant importConstant(fr.lirmm.graphik.graal.api.core.Constant constant) {
+		return makeConstant(constant.getIdentifier().toString());
+	}
+	
+	/**
+	 * Converts a {@link Set} of {@link fr.lirmm.graphik.graal.api.core.Constant Graal Constants} into a {@link Set} of {@link Constant VLog4J Constants}.
+	 * @param constants {@link Set} of {@link fr.lirmm.graphik.graal.api.core.Constant Graal Constants}
+	 * @return {@link Set} of {@link Constant VLog4J Constants}
+	 */
+	@SuppressWarnings("unused")
+	private static Set<Constant> importConstants(Set<fr.lirmm.graphik.graal.api.core.Constant> constants) {
+		Set<Constant> result = new HashSet<>();
+		for (fr.lirmm.graphik.graal.api.core.Constant constant : constants) {
+			result.add(importConstant(constant));
+		}
+		return result;
+	}
+	
+	/**
+	 * Converts a {@link fr.lirmm.graphik.graal.api.core.Predicate Graal Predicate} into a {@link Predicate VLog4J Predicate}.
+	 * @param predicate A {@link fr.lirmm.graphik.graal.api.core.Predicate Graal Predicate}
+	 * @return A {@link Predicate VLog4J Predicate}
+	 */
+	private static Predicate importPredicate(fr.lirmm.graphik.graal.api.core.Predicate predicate) {
+		return makePredicate(predicate.getIdentifier().toString(), predicate.getArity());
+	}
+
+	/**
+	 * Converts a {@link ConjunctiveQuery Graal Query} into a {@link ImportedGraalQuery}.
+	 * To use this with the {@link Reasoner}, add the {@code rule} from {@link ImportedGraalQuery#getRule()} as a Rule via {@link Reasoner#addRules(Rule...)}
+	 * and use it as the Atom for {@link Reasoner#answerQuery(Atom, boolean)}.
+	 * 
+	 * <p>
+	 * <b>WARNING</b>: The supplied {@code identifier} will be used to create a predicate containing all
+	 * answer variables from the {@code query}. If you use this identifier in another predicate, you will get conflicts.
+	 * </p>
+	 * @param identifier
+	 * @param query
+	 * @return
+	 */
+	public static ImportedGraalQuery importQuery(String identifier, ConjunctiveQuery query) {
```

Comment:
It is not clear enough what parameter identifier is used for. 
A suggestion is, instead of using object ImportedGraalQuery, have the method "conjunctiveQueryToRule(String ruleHeadPredicateName, ConjunctiveQuery conjunctiveQuery), and explain in the Javadoc that the returned rule will contain an atomic head, with given Predicate name and variables from conjunctiveQuery query variables. This rule head can be used as a query atom by the vlog4j reasoner.

Another suggestion is to use separate class, ass you do (perhaps, instead of ImportedGraalQuery, name it ConjunctiveQueryToRule), but rename field "identifier" to something more suggestive, and getQuery() to getQueryAtom()

Reply:
I would opt for the second version to avoid having to call `.getHead().getAtoms().get(0)` which, besides being verbose, relies on a magic number.

Reply:
(Hopefully) clarified the usage of `convertGraalQuery`.